addons/account/models/account_move_line.py

```diff
@@ -1567,8 +1571,7 @@ def changed(fname):
                 and 'balance' not in protected.get(line, {})
                 and (not changed('balance') or (line not in before and not line.balance))
             ):
-                balance = line.company_id.currency_id.round(line.amount_currency / line.currency_rate)
-                line.balance = balance
+                line.balance = line._get_line_balance()
```

Comment:
These fields (like, discount) are part of the line already, and the correct computation is done over and over in this same file, so I guess we should be fixing it here, instead of delegating to `purchase`.

Anyway this is not even the problem: the `purchase.order` generating this invoice should already send us the correct values for balance, instead of relying on synchronization, see the main text of the review above ☝🏻